### PR TITLE
Fix 8044 - Print name of enum passed a tmpl param

### PR DESCRIPTION
**Pull request by Geod24**
The fix was actually quite trivial. Looking at the code flow, I doubt this actually change any other diagnostic, but let's see what the CI says.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by Geod24** [comment]
> I doubt this actually change any other diagnostic

It does, but for the best!

**Comment by Geod24** [comment]
This should also fix https://issues.dlang.org/show_bug.cgi?id=14067 (marked as a duplicate now) and so improve the documentation.

**Comment by Geod24** [comment]
```
-----------------------------------------------------------
ERROR: Newly generated header file (/home/circleci/dmd/generated/linux/release/64/frontend.h) doesn't match with the reference header file (/home/circleci/dmd/src/dmd/frontend.h)

DETAILS:

diff --git a/home/circleci/dmd/src/dmd/frontend.h b/home/circleci/dmd/generated/linux/release/64/frontend.h
index a845385d4..86c99e3aa 100644
--- a/home/circleci/dmd/src/dmd/frontend.h
+++ b/home/circleci/dmd/generated/linux/release/64/frontend.h
@@ -1703,7 +1703,7 @@ struct ParameterList
     Array<Parameter*>* parameters;
     StorageClass stc;
     VarArg varargs;
-    ParameterList(Array<Parameter*>* parameters, VarArg varargs = cast(VarArg)cast(ubyte)0u, StorageClass stc = 0);
+    ParameterList(Array<Parameter*>* parameters, VarArg varargs = VarArg.none, StorageClass stc = 0);
     size_t length();
     Parameter* opIndex(size_t i);
     ParameterList() :

===============
```

The header generation depended on that bug ? :|

File: src/dmd/frontend.h [hunk 1]
```diff
@@ -1703,7 +1703,7 @@ struct ParameterList
     Array<Parameter*>* parameters;
     StorageClass stc;
     VarArg varargs;
-    ParameterList(Array<Parameter*>* parameters, VarArg varargs = cast(VarArg)cast(ubyte)0u, StorageClass stc = 0);
+    ParameterList(Array<Parameter*>* parameters, VarArg varargs = VarArg.none, StorageClass stc = 0);
```
Comment (ibuclaw) [inline src/dmd/frontend.h:1706]: This isn't valid C++, it's `VarArg::none`.

**Comment by ibuclaw** [comment]
> The header generation depended on that bug ? :|

Strictly for dmd, I don't think we require any of these automatic constructors, as initialisation is handed over to the D layer.

**Comment by UplinkCoder** [comment]
If I recall correctly there are still cases in which you'll see `cast(EnumType)value`
and the comment out `assert(0)` was there for that reason, to help debug such cases.

File: src/dmd/frontend.h (hunk 1 above)
Comment (Geod24) [reply src/dmd/frontend.h:1706]: CC @MoonlightSentinel

**Comment by Geod24** [comment]
> If I recall correctly there are still cases in which you'll see `cast(EnumType)value`

If you find any, please report. As can be seen from the diff in the testsuite, many, if not all of them, should disappear. I left the `cast` in to be sure no regression is introduced.

> and the comment out assert(0) was there for that reason, to help debug such cases.

If someone want to debug such case, adding back and `assert` or `printf` is trivial. There's no need to have this `assert` here as it's not something frequently needed as opposed to the other debugging printf (e.g. `printf` in semantic routines).

File: src/dmd/frontend.h (hunk 1 above)
Comment (MoonlightSentinel) [reply src/dmd/frontend.h:1706]: Access to enum/static members needs more work but that's unrelated to this PR. (`frontend.h` doesn't compile anyways)

**Review by MoonlightSentinel — APPROVED** [review]

**Comment by MoonlightSentinel** [comment]
Reduced test for the `dtoh` segfault:
```d
enum E { m }

extern (C++) void foo(ulong = E.m) {}
```
Working on it.

File: src/dmd/dtoh.d [hunk 2]
```diff
@@ -1772,7 +1772,8 @@ public:
         ident = null;
         version (all)
         {
-            if (p.defaultArg && p.defaultArg.op >= TOK.int32Literal && p.defaultArg.op < TOK.struct_)
+            if (p.defaultArg && !p.defaultArg.type.isTypeEnum() &&
```
Comment (MoonlightSentinel) [inline src/dmd/dtoh.d:1775]: `defaultArg.type` is apparently null for enum members as default values....

**Comment by Geod24** [comment]
This should be good to ggo

**Pull request merged**